File: ml-service/utils/logger.py
import logging
import json
import datetime
import os

class JsonFormatter(logging.Formatter):
    def format(self, record):
        log_record = {
            "timestamp": datetime.datetime.fromtimestamp(record.created).isoformat(),
            "level": record.levelname,
            "name": record.name,
            "message": record.getMessage(),
            "module": record.module,
            "funcName": record.funcName,
            "lineno": record.lineno,
        }
        if record.exc_info:
            log_record["exc_info"] = self.formatException(record.exc_info)
        return json.dumps(log_record)

def setup_logger(name="krishimitraai", level=logging.INFO):
    logger = logging.getLogger(name)
    logger.setLevel(level)
    
    # Prevent duplicate handlers
    if not logger.handlers:
        handler = logging.StreamHandler()
        
        # Use JSON formatter in production
        if os.environ.get("ENVIRONMENT") == "production":
            handler.setFormatter(JsonFormatter())
        else:
            handler.setFormatter(logging.Formatter("[%(asctime)s] %(levelname)s [%(name)s.%(funcName)s:%(lineno)d] %(message)s"))
            
        logger.addHandler(handler)
        
    return logger

File: backend/utils/logger.js
import winston from 'winston';

const { combine, timestamp, json, printf, colorize, align } = winston.format;

export const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || 'info',
  format: combine(
    timestamp({
      format: 'YYYY-MM-DD HH:mm:ss',
    }),
    process.env.NODE_ENV === 'production' ? json() : combine(
      colorize(),
      align(),
      printf((info) => `[${info.timestamp}] ${info.level}: ${info.message}`)
    )
  ),
  transports: [
    new winston.transports.Console(),
  ],
});
